File: src/main/java/novemberizing/ds/Executable.java
package novemberizing.ds;

/**
 *
 * @author novemberizing, me@novemberizing.net
 * @since 2017. 1. 17.
 */
@SuppressWarnings("unused")
public interface Executable {
    void execute(Executor executor);
    boolean completed();
}

File: src/main/c++/novemberizing/ds/executable.cc
#include "executable.hh"

namespace novemberizing { namespace ds {

Executable::Executable(void) : __executor(nullptr)
{
    FUNCTION_START("");
    FUNCTION_END("");
}

Executable::~Executable(void)
{
    FUNCTION_START("");

    if(__executor!=nullptr)
    {
        Executor * executor = __executor;
        executor->completed(this);
        __executor = nullptr;
    }

    FUNCTION_END("");
}

void Executable::execute(Executor * executor)
{
    FUNCTION_START("");
    if(__executor!=nullptr)
    {
        WARNING_LOG("");
    }
    __executor = executor;
    execute();
    FUNCTION_END("");
}

} }
